deploy/buildtools/bitbuilder.py

```diff
@@ -0,0 +1,320 @@
+import abc
+import yaml
+import json
+import time
+import random
+import string
+import logging
+import os
+from fabric.api import prefix, local, run, env, lcd, parallel # type: ignore
+from fabric.contrib.console import confirm # type: ignore
+from fabric.contrib.project import rsync_project # type: ignore
+
+from awstools.afitools import *
+from awstools.awstools import send_firesim_notification
+from util.streamlogger import StreamLogger, InfoStreamLogger
+
+from typing import Optional, TYPE_CHECKING
+# TODO: Solved by "from __future__ import annotations" (see https://stackoverflow.com/questions/33837918/type-hints-solve-circular-dependency)
+if TYPE_CHECKING:
+    from buildtools.buildconfig import BuildConfig
+else:
+    BuildConfig = object
+
+rootLogger = logging.getLogger()
+
+def get_deploy_dir() -> str:
+    """Determine where the firesim/deploy directory is and return its path.
+
+    Returns:
+        Path to firesim/deploy directory.
+    """
+    with StreamLogger('stdout'), StreamLogger('stderr'):
+        deploydir = local("pwd", capture=True)
+    return deploydir
+
+class BitBuilder(metaclass=abc.ABCMeta):
+    """Abstract class to manage how to build a bitstream for a build config.
+
+    Attributes:
+        build_config: Build config to build a bitstream for.
+    """
+    build_config: BuildConfig
+
+    def __init__(self, build_config: BuildConfig) -> None:
+        """
+        Args:
+            build_config: Build config to build a bitstream for.
+        """
+        self.build_config = build_config
+
+    @abc.abstractmethod
+    def replace_rtl(self) -> None:
```

Comment:
Can we somehow assert that this has to run on localhost? If this is too complicated, I think we should mention what functions in this class are designed to run locally vs anywhere